core/trino-main/src/main/java/io/trino/memory/ClusterMemoryManager.java

```diff
@@ -225,13 +240,9 @@ public synchronized void process(Iterable<QueryExecution> runningQueries, Suppli
         if (!(lowMemoryKiller instanceof NoneLowMemoryKiller) &&
                 outOfMemory &&
                 !queryKilled &&
-                nanosSince(lastTimeNotOutOfMemory).compareTo(killOnOutOfMemoryDelay) > 0) {
-            if (isLastKillTargetGone(runningQueries)) {
-                callOomKiller(runningQueries);
-            }
-            else {
-                log.debug("Last killed target is still not gone: %s", lastKillTarget);
-            }
+                nanosSince(lastTimeNotOutOfMemory).compareTo(killOnOutOfMemoryDelay) > 0 &&
+                lastKillTarget.isEmpty() && nanosSince(lastTimeKillCompleted).compareTo(killOnOutOfMemoryAfterKillDelay) > 0) {
+            callOomKiller(runningQueries);
```

Comment:
The sequence of operations in the `process` method looks very weird.

I would rather expect it to be something like this

```
// update memory reservation on each node
updateNodes();

// update memory pool
updateMemoryPool(Iterables.size(runningQueries));

... logic related to enforcing limits and freeing up memory pools ...
```

Do you think it might be the reason of these weird inconsistencies?

Reply:
I think it does not matter; `updateNodes()` is asynchronous anyway.